source/LiteralModel.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <mariana-trench/EventLogger.h>
+#include <mariana-trench/LiteralModel.h>
+#include <mariana-trench/Positions.h>
+
+namespace marianatrench {
+
+namespace {
+
+class LiteralModelConsistencyError {
+ public:
+  static void raise(const std::string& what) {
+    ERROR(1, "Literal Model Consistency Error: {}", what);
+    EventLogger::log_event("regex_model_consistency_error", what);
+  }
+};
+
+}
+
+static void check_taint_config_consistency(const std::string& pattern, const TaintConfig& frame) {
+  if (frame.kind() == nullptr) {
+    LiteralModelConsistencyError::raise(fmt::format(
+        "Model for regex `{}` must have a source kind.", pattern));
+  }
+}
+
+static Taint create_taint(const std::string& pattern, const std::vector<TaintConfig>& sources) {
+  Taint all_sources;
+  for (const auto& source : sources) {
+    mt_assert(source.is_leaf());
+    check_taint_config_consistency(pattern, source);
+    all_sources.join_with(Taint{std::move(source)});
+  }
+  return all_sources;
+}
+
+LiteralModel::LiteralModel() : regex_(pattern_) { }
+
+LiteralModel::LiteralModel(
+    std::string&& pattern,
```

Comment:
We usually take by copy in those cases.
See https://www.youtube.com/watch?v=PNRju6_yn3o

Also, as a rule of thumb, let's be consistent with the rest of the codebase.

Reply:
Well technically `std::string_view` is worse in the case where we move a `std::string`, it will make an extra copy.
but performance is not a problem here unless proven otherwise.

Reply:
I am not sure I am seeing an extra copy in that case? `std::string_view` doesn't have a constructor for `std::string &&`, so won't it just do the same as usual and just copy the pointer and length? And the constructor of `RE2` will always copy exactly once, irrespective of whether we use `const std::string &` or `const StringPiece &`/`std::string_view`, since `StringPiece` has a `string_view` constructor.

I thus defaulted to `std::string_view` since it doesn't require the caller to use `std::move`. Having said that, we invoke this constructor exactly once with an `rvalue` anyway, so I just changed it to an `std::string` to keep things simple and consistent. 🙂